cli/src/pcluster/models/common.py

```diff
@@ -239,3 +252,413 @@ def dump_json(self):
+
+    @property
+    def artifact_directory(self):
+        """Return artifact directory."""
+        if self.__artifact_directory is None:
+            self.__artifact_directory = self._get_artifact_dir()
+        return self.__artifact_directory
+
+    # --------------------------------------- S3 bucket utils --------------------------------------- #
+
+    def get_bucket_name(self):
+        """
+        Get ParallelCluster bucket name.
+
+        :param account_id
+        :param region
+        :return ParallelCluster bucket name e.g. parallelcluster-b9033160b61390ef-v1-do-not-delete
+        """
+        return "-".join(
+            [
+                "parallelcluster",
+                self._generate_s3_bucket_hash_suffix(),
+                PCLUSTER_S3_BUCKET_VERSION,
+                "do",
+                "not",
+                "delete",
+            ]
+        )
+
+    def check_bucket_exists(self):
+        """Check bucket existence by bucket name."""
+        AWSApi.instance().s3.head_bucket(bucket_name=self.name)
+
+    def create_bucket(self):
+        """
+        Create a new S3 bucket.
+
+        :param bucket_name: name of the S3 bucket to create
+        :param region: aws region
+        :raise AWSClientError if bucket creation fails
+        """
+        if self.region != "us-east-1":
+            AWSApi.instance().s3.create_bucket(bucket_name=self.name, configuration={"LocationConstraint": self.region})
+        else:
+            AWSApi.instance().s3.create_bucket(bucket_name=self.name)
+        try:
+            self._configure_s3_bucket()
+        except AWSClientError as e:
+            if "NoSuchBucket" not in str(e):
+                LOGGER.error("Failed when configuring cluster S3 bucket.")
+                self.delete_bucket()
+                raise
+
+    def delete_bucket(self):
+        """
+        Delete an S3 bucket together with all stored objects.
+
+        :param bucket_name: name of the S3 bucket to delete
+        """
+        try:
+            if not self._is_custom_bucket:
+                LOGGER.info("Deleting bucket %s", self.name)
+                AWSApi.instance().s3_resource.delete_all_objects(bucket_name=self.name)
+                AWSApi.instance().s3_resource.delete_all_object_versions(bucket_name=self.name)
+                AWSApi.instance().s3_resource.delete_bucket(bucket_name=self.name)
+        except AWSClientError as e:
+            if "NoSuchBucket" not in str(e):
+                LOGGER.warning(
+                    "Failed to delete S3 bucket %s. Please delete it manually.",
+                    self.name,
+                )
+                raise
+
+    # --------------------------------------- S3 objects utils --------------------------------------- #
+
+    def get_object_key(self, object_type, object_name):
+        """Get object key of an artifact."""
+        return "/".join([self.artifact_directory, object_type, object_name])
+
+    def generate_artifact_dir(self):
```

Comment:
I think generate_artifact_dir can't be called with constructor, because it calls random string generation function, this function should only be called in create command. Otherwise, we will generate different artifacts_dir for same image/cluster.

The logic here is:
If the consumer is pcluster command or pcluster build-image, call generate_artifact_dir method, and assign to the parameter artifact_directory.
If the consumer is other commands, call get_artifact_dir to retrieve the artifact_directory from stack or image, and assign to the parameter artifact_directory.